configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java

```diff
@@ -394,13 +403,84 @@ private void loadSessionIfActive(RemoteSession session) {
         for (ApplicationId applicationId : applicationRepo.activeApplications()) {
             if (applicationRepo.requireActiveSessionOf(applicationId) == session.getSessionId()) {
                 log.log(Level.FINE, () -> "Found active application for session " + session.getSessionId() + " , loading it");
-                applicationRepo.reloadConfig(session.ensureApplicationLoaded());
+                applicationRepo.reloadConfig(ensureApplicationLoaded(session));
                 log.log(Level.INFO, session.logPre() + "Application activated successfully: " + applicationId + " (generation " + session.getSessionId() + ")");
                 return;
             }
         }
     }
 
+    void prepareRemoteSession(RemoteSession session) {
+        SessionZooKeeperClient sessionZooKeeperClient = createSessionZooKeeperClient(session.getSessionId());
+        Curator.CompletionWaiter waiter = sessionZooKeeperClient.getPrepareWaiter();
+        ensureApplicationLoaded(session);
+        notifyCompletion(waiter, session);
+    }
+
+    public synchronized ApplicationSet ensureApplicationLoaded(RemoteSession session) {
+        Optional<ApplicationSet> applicationSet = session.applicationSet();
+        if (applicationSet.isPresent()) {
+            return applicationSet.get();
+        }
+
+        ApplicationSet newApplicationSet = loadApplication(session);
+        RemoteSession newSession = new RemoteSession(session.getTenantName(),
+                                    session.getSessionId(),
+                                    session.getSessionZooKeeperClient(),
+                                    Optional.of(newApplicationSet));
+        remoteSessionCache.addSession(newSession);
+        return newApplicationSet;
+    }
+
+    void confirmUpload(RemoteSession session) {
+        Curator.CompletionWaiter waiter = session.getSessionZooKeeperClient().getUploadWaiter();
```

Comment:
Why `session.getSessionZooKeeperClient()` here, and `createSessionZooKeeperClient(session.getSessionId())` elsewhere?

Reply:
Yes, both are used, I want to clean up that, will do in a later PR